File: choco3/PerfectSquareSequence.java
/**
 *
 * Perfect square sequence in Choco3
 *
 * From 'Fun with num3ers'
 * "Sequence"
 * http://benvitale-funwithnum3ers.blogspot.com/2010/11/sequence.html
 * """
 * If we take the numbers from 1 to 15 
 *    (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15) 
 * and rearrange them in such an order that any two consecutive 
 * numbers in the sequence add up to a perfect square, we get,
 *  
 *  8     1     15     10     6     3     13     12      4      5     11     14        2      7      9
 *      9    16    25     16     9     16     25     16     9     16     25     16       9     16
 *
 *  
 * I ask the readers the following:
 *  
 * Can you take the numbers from 1 to 25 to produce such an arrangement?
 * How about the numbers from 1 to 100?
 * """
 *
 * Via http://wildaboutmath.com/2010/11/26/wild-about-math-bloggers-111910
 *
 *
 * Choco3 model by Hakan Kjellerstrand (hakank@gmail.com)
 * http://www.hakank.org/choco3/
 *
 */
import org.kohsuke.args4j.Option;
import org.slf4j.LoggerFactory;
import samples.AbstractProblem;
import solver.ResolutionPolicy;
import solver.Solver;
import solver.constraints.Constraint;
import solver.constraints.IntConstraintFactory;
import solver.search.strategy.IntStrategyFactory;
import solver.search.strategy.selectors.values.InDomainMax;
import solver.search.strategy.selectors.values.InDomainMiddle;
import solver.search.strategy.selectors.values.InDomainMin;
import solver.search.strategy.selectors.values.InDomainRandom;
import solver.variables.IntVar;
import solver.variables.BoolVar;
import solver.variables.VariableFactory;
import solver.explanations.ExplanationFactory;
import solver.search.strategy.strategy.AbstractStrategy;
import solver.search.strategy.selectors.variables.*;
import solver.search.strategy.strategy.Assignment;
import util.ESat;
import util.tools.ArrayUtils;


public class PerfectSquareSequence extends AbstractProblem {

  @Option(name = "-n", usage = "Size of sequence (default 15).", required = false)
  int n = 15;

  @Option(name = "-solutions", usage = "Number of solutions to show (default 1).", required = false)
  int solutions = 1;


  IntVar[] x;

  @Override
  public void buildModel() {

    int squares[] = new int[n-1];
    for(int i = 1; i < n; i++) {
      squares[i-1] = i*i;
    }

    x = VariableFactory.enumeratedArray("x", n, 1, n, solver);

    solver.post(IntConstraintFactory.alldifferent(x, "BC"));

    for(int i = 1; i < n; i++) {
      IntVar s = VariableFactory.bounded("s_"+i, 0, n*n, solver);
      solver.post(IntConstraintFactory.sum(new IntVar[]{x[i-1],x[i]},s));
      solver.post(IntConstraintFactory.member(s,squares));
    }

    // symmetry breaking
    solver.post(IntConstraintFactory.arithm(x[0], "<", x[n-1]));



  }

  @Override
  public void createSolver() {
    solver = new Solver("PerfectSquareSequence");
  }

  @Override
  public void configureSearch() {
    solver.set(IntStrategyFactory.firstFail_InDomainMin(x));
  }

  @Override
  public void solve() {
    // solver.findOptimalSolution(ResolutionPolicy.MINIMIZE, z);
    solver.findSolution();

  }

  @Override
  public void prettyOut() {

    if(solver.isFeasible() == ESat.TRUE) {
      int num_solutions = 0;
      do {
        
        for(int i = 0; i < n; i++) {
          System.out.print(x[i].getValue() + " ");
        }
        System.out.println();

        num_solutions++;

        if (solutions > 0 && num_solutions >= solutions) {
          break;
        }

      } while (solver.nextSolution() == Boolean.TRUE);

      System.out.println("It was " + num_solutions + " solutions.");

    } else {

      System.out.println("Problem is not feasible.");

    }

  }


  //
  // main
  //
  public static void main(String args[]) {

    new PerfectSquareSequence().execute(args);

  }
}

File: google_or_tools/perfect_square_sequence.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License"); 
# you may not use this file except in compliance with the License. 
# You may obtain a copy of the License at 
#
#     http://www.apache.org/licenses/LICENSE-2.0 
#
# Unless required by applicable law or agreed to in writing, software 
# distributed under the License is distributed on an "AS IS" BASIS, 
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
# See the License for the specific language governing permissions and 
# limitations under the License. 

"""

  Perfect square sequence in Google CP Solver.

  From 'Fun with num3ers'
  'Sequence'
  http://benvitale-funwithnum3ers.blogspot.com/2010/11/sequence.html
  '''
  If we take the numbers from 1 to 15 
      (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15) 
  and rearrange them in such an order that any two consecutive 
  numbers in the sequence add up to a perfect square, we get,
  
  8     1     15     10     6     3     13     12      4      5     11     14        2      7      9
      9    16    25     16     9     16     25     16     9     16     25     16       9     16
  
  
  I ask the readers the following:
  
  Can you take the numbers from 1 to 25 to produce such an arrangement?
  How about the numbers from 1 to 100?
  '''

  Via http://wildaboutmath.com/2010/11/26/wild-about-math-bloggers-111910

  Compare with the following model:
  * MiniZinc: http://www.hakank.org/minizinc/perfect_square_sequence.mzn

  This model was created by Hakan Kjellerstrand (hakank@gmail.com)
  Also see http://www.hakank.org/google_or_tools/
  
"""
import sys
import math

from ortools.constraint_solver import pywrapcp

def is_square(v):
    solver = v.solver()
    ub = int(math.sqrt(v.Max()))
    z = solver.IntVar(1, ub, 'z')
    solver.Add(z*z == v)

def main(n=15, print_solutions=True, show_num_sols=0):

    sys.stdout.flush()
    
    # Create the solver.
    solver = pywrapcp.Solver('17x17 challenge')

    #
    # data
    #
    print 'n: ', n

    # create the table of possible squares
    squares = []
    for i in range(1, int(math.sqrt(n*n))):
        squares.append(i*i)
    # print "squares:", squares, len(squares)

    # print 'valid squares:', squares

    # declare variables
    x = [solver.IntVar(1, n, 'x[%i]'%i) for i in range(n)]

    #
    # constraints
    #
    
    solver.Add(solver.AllDifferent(x, True))
    for i in range(1, n):
        # solver.Add(solver.Square(x[i-1]+x[i]))
        # is_square(x[i-1]+x[i])
        solver.Add(solver.MemberCt(x[i-1]+x[i], squares))


    # symmetry breaking
    solver.Add(x[0] < x[n-1])

    #
    # solution and search
    #
    db = solver.Phase(x,
                      # solver.CHOOSE_FIRST_UNBOUND,
                      solver.INT_VAR_DEFAULT,                      
                      solver.INT_VALUE_DEFAULT)

    solver.NewSearch(db)
    
    num_solutions = 0
    while solver.NextSolution():
        num_solutions += 1
        if print_solutions:
            x_val = [x[i].Value() for i in range(n)]
            print 'x: ', x_val
            # for i in range(1, n):
            #    print (x_val[i-1]+x_val[i]),
            # print
        if show_num_sols > 0 and num_solutions >= show_num_sols:
            break

    solver.EndSearch()
    
    print
    print "num_solutions:", num_solutions
    print
    if print_solutions:
        print "failures:", solver.Failures()
        print "branches:", solver.Branches()
        print "wall_time:", solver.WallTime()

    return num_solutions

n = 25
if __name__ == '__main__':
    if len(sys.argv) > 1:
        n = int(sys.argv[1])

    if n == 0:
        sols = []
        for i in range(1,100):
           num = main(i, False, 0)
           sols.append((i,num))
        print sols
    else:
        main(n, True, 0)
